tls/cfsslscan_handshake.go

```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
 	if err != nil {
 		return
 	}
+	if CipherSuites[serverHello.cipherSuite].EllipticCurve {
+		// Prime the connection, if necessary, for key
+		// exchange messages by reading off the certificate
+		// message and, if necessary, the OCSP stapling
+		// message
+		var msg interface{}
+		msg, err = c.readHandshake()
+		if err != nil {
+			return
+		}
+		certMsg, ok := msg.(*certificateMsg)
+		if !ok || len(certMsg.certificates) == 0 {
+			c.sendAlert(alertUnexpectedMessage)
+			err = unexpectedMessageError(certMsg, msg)
+			return
+		}
+		if serverHello.ocspStapling {
+			msg, err = c.readHandshake()
+			if err != nil {
+				return
+			}
+			_, ok := msg.(*certificateStatusMsg)
+			if !ok {
+				c.sendAlert(alertUnexpectedMessage)
+				return
+			}
+		}
+
+		var skx *serverKeyExchangeMsg
+		skx, err = c.exchangeKeys()
+		if skx.raw[0] != typeServerKeyExchange {
+			err = errors.New("tls: received something other than the expected ServerKeyExchange.")
+			return
+		}
+		if err != nil {
+			return
+		}
+		if len(skx.key) < 4 {
+			err = errServerKeyExchange
+			return
+		}
+		curveType = uint16(skx.key[0])
+		// If we have a named curve, report which one it is.
+		if curveType == 3 {
+			curveID = CurveID(skx.key[1])<<8 | CurveID(skx.key[2])
+		}
+	}
 	cipherID, version = serverHello.cipherSuite, serverHello.vers
 	return
-
 }
 
 // sayHello is the backend to SayHello that returns a full serverHelloMsg for processing.
```

Comment:
I see it on e.g. cloudflare.com. I also guessed it had something to do with large clientHello messages, but I don't think that's correct - it doesn't happen for every request. And the connections don't hang - they send a TLS "Alert" message. So perhaps it's really happening when an ECDH cipher suite is requested but no requested curves are available.

From RFC 5246:
> handshake_failure
>      Reception of a handshake_failure alert message indicates that the
>      sender was unable to negotiate an acceptable set of security
>      parameters given the options available.  This is a fatal error.

I've went throught and characterized the errors pretty well: there's exactly one of these messages for each `(serverCipher, TLSVersion)` pair negotiated by the server (where `tls.CipherSuites[serverCipher].EllipticCurve == true`), which suggests to me that this is happening in response to the server actually not being able to negotiate anything (which is odd, because the `clientHello` corresponding to the `Alert` message actually contains a negotiable set of parameters, namely all the named curves, at least one of which should be valid). This led me to wonder if you're not allowed to request the curveID values that correspond to non-named curves, and you aren't, per http://tools.ietf.org/html/rfc4492#section-5.4. Dropping those values from the initial set of `tls.CurveID`s presented didn't change where the errors occur, though.

There is also 1 additional error per TLSVersion, which has _most_ of the CipherSuites (315 out of 320) and all of the curve options.